manim/utils/hashing.py

```diff
@@ -1,20 +1,140 @@
+    ALREADY_PROCESSED_PLACEHOLDER = "ALREADY PROCESSED"
+
+    @classmethod
+    def reset_already_processed(cls):
+        cls._already_processed.clear()
+
+    @classmethod
+    def check_already_processed_decorator(cls: "_Memoizer", is_method=False):
+        """Decorator to handle the arguments that goes through the decorated function. Returns _ALREADY_PROCESSED_PLACEHOLDER if the obj has been processed, or lets the decorated function call go ahead.
+
+        Parameters
+        ----------
+        is_method : bool, optional
+            Whether the function passed is a method, by default Falsep
+        """
+
+        def layer(func):
+            # NOTE : There is probably a better to separate both case when func is a method or a function.
+            if is_method:
+                return lambda self, obj: cls._handle_already_processed(
+                    obj, default_function=lambda obj: func(self, obj)
+                )
+            return lambda obj: cls._handle_already_processed(obj, default_function=func)
+
+        return layer
+
+    @classmethod
+    def check_already_processed(cls, obj: Any) -> Any:
+        """Checks if obj has been already processed, returns itself if it hasn't been or the value of _ALREADY_PROCESSED_PLACEHOLDER if it has not.
+        Marks the object as processed in the second case.
+
+        Parameters
+        ----------
+        obj : Any
+            The object to check.
+
+        Returns
+        -------
+        Any
+            Either the object itself or the placeholder.
+        """
+        # When the object is not memorized, we return the object itself.
+        return cls._handle_already_processed(obj, lambda x: x)
+
+    @classmethod
+    def mark_as_processed(cls, obj: Any) -> None:
+        """Marks an object as processed.
+
+        Parameters
+        ----------
+        obj : Any
+            The object to mark as processed.
+        """
+        cls._handle_already_processed(obj, lambda x: x)
+        return cls._return(obj, id, lambda x: x, memoizing=False)
+
+    @classmethod
+    def _handle_already_processed(
+        cls, obj, default_function: typing.Callable[[Any], Any]
+    ):
+        if (
+            isinstance(
+                obj,
+                (
+                    int,
+                    float,
+                    str,
+                    complex,
+                ),
+            )
+            and obj not in [None, cls.ALREADY_PROCESSED_PLACEHOLDER]
+        ):
+            # It makes no sense (and it'd slower) to memoize objects of these primitive types.
+            # Hence, we simply return the object.
+            return obj
+        if isinstance(obj, collections.abc.Hashable):
+            try:
+                return cls._return(obj, hash, default_function)
+            except TypeError:
+                # In case of an error with the hash (eg an object is marked as hashable but contains a non hashable within it)
```

Comment:
The main goal and problem with scene caching is to be flexible enough for anything the user could plug in. I don't think it would make sense to throw an exception here. 

`id` refers to the python's built-in function, that returns the address of the object.

The idea is that hash is preferred to be used for membership test, since if two objects have the same _content_ they have to be -when possible- hashed by scene caching only once, and the second one should be ignored.
 This content test equality is done with hashing, and when the object is not hashable for some reason equality by identity is used, which still work but is less "broad" than working with hash. That's my reasoning.

Reply:
In what kinds of situations would there be an object marked as hashable but containing a non-hashable entry within it? (I guess I don't know what 'object' and 'containing' mean in this case - is it Python objects and containing being members and hashable meaning Python hashable? or is it something else?)

Would a warning or a info be appropriate here, in case other people have a "why isn't this caching" kind of question? Idk, not sure.

Reply:
> In what kinds of situations would there be an object marked as hashable but containing a non-hashable entry within it?

When an immutable object contains a mutable object within it (for example, a tuple of a list). 
But after rechecking, it appears that checking with `isinstance(obj, collections.abc.Hashable)` recursively check for hashablity of all the sub elements, so you're right.

Concerning the warning, I'm strongly against because it goes against the principle of scene caching. Scene caching should be flexible enough to allow the user to plug more or less anything, and it will still work without any error or warning. I think that scene caching should adapt to the user, not the reverse. 
On another note, it would be impossible to do otherwise for the moment, because manim objects have not been designed to be serialized (too many attributes, etc), so the only solution is to build something very flexible.

Reply:
I think the "don't report any problem" can lead to weird behavior on when things are cached vs not... that might be more frustrating than the warnings but idk. I follow your reasoning though.

Reply:
> can lead to weird behavior on when things are cached vs not... that might be more frustrating than the warnings but idk.

What do you mean ?